### PR TITLE
Add documentation for `#[MapUploadedFile]` attribute

### LINKED ISSUE #19802: [HttpKernel] Introduce `#[MapUploadedFile]` controller argument attribu…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#49978
| PR author(s) | @renedelima
| Merged in    | 7.1

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 7.1 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smiley:

### PR DESCRIPTION
**Pull request by renedelima**
Closes #19802

### CODE REVIEW

File: controller.rst [hunk 1]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
 
     The ``type`` option of ``#[MapRequestPayload]`` was introduced in Symfony 7.1.
 
+Mapping Uploaded File
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can map ``Uploaded Files`` to the controller arguments and add ``constraints``.
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
You can map ``UploadedFile``s to the controller arguments and add ``constraints``.
```

File: controller.rst [hunk 2]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+By default the argument resolver will try to match the argument name with the submitted file name.
+
+The constraints defined by you are performed before injecting the uploaded file into the controller argument.
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
The constraints defined by you are performed before injecting the `UploadedFile` into the controller argument.
```

File: controller.rst [hunk 3]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+    {
+        public function _invoke(
+            #[MapUploadedFile( new Assert\File(mimeTypes: ['image/png', 'image/jpeg']) )]
+            ?UploadedFile $picture
+        ): Response {
+            // ...
+        }
+    }
+
+You can also map a collection of uploaded files to array or variadic arguments::
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
You can also map a collection of `UploadedFile`s::
```

File: controller.rst [hunk 4]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+
+    #[MapUploadedFile( new Assert\File(mimeTypes: ['application/pdf']) )]
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
    #[MapUploadedFile(new Assert\File(mimeTypes: ['application/pdf']))]
```

File: controller.rst [hunk 5]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+    array $documents
+
+    #[MapUploadedFile( new Assert\File(mimeTypes: ['application/pdf']) )]
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
    #[MapUploadedFile(new Assert\File(mimeTypes: ['application/pdf']))]
```

File: controller.rst [hunk 6]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+Mapping Uploaded File
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can map ``Uploaded Files`` to the controller arguments and add ``constraints``.
+By default the argument resolver will try to match the argument name with the submitted file name.
+
+The constraints defined by you are performed before injecting the uploaded file into the controller argument.
+An ``Unprocessable Content HTTP``` exception is thrown if a constraint violation is detected.
+
+    #[Route('/user/picture', methods: ['PUT'])]
```
Comment (OskarStark) [inline controller.rst:608]: Please add the use statements of the UploadedFile, the Assert and the MapUploadedFile

File: controller.rst [hunk 7]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+You can also map a collection of uploaded files to array or variadic arguments::
+
+    #[MapUploadedFile( new Assert\File(mimeTypes: ['application/pdf']) )]
+    array $documents
+
+    #[MapUploadedFile( new Assert\File(mimeTypes: ['application/pdf']) )]
+    UploadedFile ...$documents
+
+Further, you can match the argument with a custom name,
+and also change the ``HTTP Status Code`` thrown when a constraint violation are detected.
```
Comment (OskarStark) [inline controller.rst]:
```suggestion
and also change the thrown HTTP status code, when constraint violations are detected::
```

File: controller.rst [hunk 8]
```diff
@@ -587,6 +587,44 @@ using the ``type`` option of the attribute::
...
+
+    #[MapUploadedFile(
+        constraints: new Assert\File(mimeTypes: ['application/pdf']),
+        name: 'something-else',
+        validationFailedStatusCode: Response::HTTP_BAD_REQUEST
+    )]
+    UploadedFile $document
+
```
Comment (OskarStark) [inline controller.rst:678]: please add a `versionadded` directive
```suggestion

.. versionadded:: 7.1

    The `#[MapUploadedFile]` attribute was introduced in Symfony 7.1.

```

**Comment by renedelima** [comment]
@OskarStark I applied your suggestions, fixed the `Code blocks` check, and made some other changes. The PR is now ready for review.

File: controller.rst [hunk 9]
```diff
@@ -587,6 +587,94 @@ using the ``type`` option of the attribute::
...
+.. tip::
+
+    The bound ``Constraints`` are performed before injecting the ``UploadedFile`` into the controller argument.
+    When a constraint violation is detected an ``HTTPException`` is thrown and the controller's
+    action is not executed.
+
+Mapping ``UploadedFile``s with no custom settings.
+
+.. code-block:: php-attributes
```
Comment (alexandre-daubois) [inline controller.rst:630]: Are these directives really needed, as there is only one code block and not tabs?

**Comment by mttsch** [comment]
Should the attribute also be added to https://symfony.com/doc/7.1/reference/attributes.html#httpkernel?

File: controller.rst (hunk 9 above)
Comment (renedelima) [reply controller.rst:630]: I had to explicitly set `php-attributes` markup format to these code blocks to pass the checkers.

**Comment by renedelima** [comment]
> Should the attribute also be added to https://symfony.com/doc/7.1/reference/attributes.html#httpkernel?

The link was added.

**Comment by javiereguiluz** [comment]
Renê, thanks a lot for adding these docs ... and congrats on your first Symfony Docs contribution 🎉 

What a beautiful contribution you did to Symfony code (and docs too!). Tomorrow I'll publish a "New in Symfony 7.1" post dedicated to this.

Note: while merging I did some tweaks and rewords. See https://github.com/symfony/symfony-docs/commit/846cbae795919e74ec57ee9d1010cb3ff16d4041. It looks like I changed a ton of things, but it's not true:

* I changed the order to display first the simple example with no constraints
* I changed the style of the controllers to show a standard controller extending AbstractController instead of an invokable controller (your original example is perfectly correct, but this article is one of the basic chapters of Symfony Docs, so I prefer to keep the same basic style as in the rest of the page)
* Other minor tweaks

So, most of your contribution stays. Thanks!

**Pull request merged**